connectors/utils.py

```diff
@@ -300,3 +300,36 @@ async def _wait_for_room(self, item):
     async def put(self, item):
         item_size = await self._wait_for_room(item)
         return await super().put((item_size, item))
+
+
+class ConcurrentTasks:
+    def __init__(self, max_concurrency=5, results_callback=None):
+        self.max_concurrency = max_concurrency
+        self.tasks = []
+        self.results_callback = results_callback
+        self._task_over = asyncio.Event()
+
+    def __len__(self):
+        return len(self.tasks)
+
+    def _callback(self, task):
+        self.tasks.remove(task)
+        self._task_over.set()
+        if self.results_callback is not None:
+            if task.exception():
+                raise fut.exception()
+            self.results_callback(task.result())
```

Comment:
I remembered I submitted this comment but I don't know where it is now

I assume the purpose of `results_callback` is to return the result if a task is supposed to return something. 

If the above assumption is true, I don't think it should be passed in `__init__`, but should be a param of `put` function, with default value of `None`. Because task A can return something but task B may not.

Reply:
What I meant was, we shouldn't add `results_callback` to `__init__`, and allow `None`.

Instead, we can make it a param of `put` function:
```python
def _callback(self, task, result=None):
    ...
    if result is not None:
        if task.exception():
            raise task.exception()
        result = task.result()


async def put(self, coroutine, result=None):
    ...
    task.add_done_callback(self._callback(result))
    return task
```

Reply:
oh, I see what you mean. I see two use cases:

1. treat all results in the same callback
2. treat each result independently

1. is the most common use case as far as I know because in general, we send the same coroutine with different arguments and the collected results are homogeneous.

For 2. the fact that `put` returns the task allows you to call await on it and gets it `result()` directly, but you will need to do it in async code.

I can add both, it does not complexity the code -- we'll see if it's useful long term